OpenShiftClientProducer reuses the KubernetesClient default bean

manusa opened the pull request:
Fix #35383

Instead of creating a new bean (which might lead to an application having two separate `KubernetesClient` instances) the `OpenShiftClientProducer` now relies on the `KubernetesClientProducer` default bean and adapts it.

To avoid circular dependency and multiple bean candidate issue I've used a `@Named` annotation to qualify the default `KubernetesClient` bean.

However, this might be problematic in the scenario where a user provides a different base `KubernetesClient` bean. In this case, unless that bean is qualified too, it won't be reused. This is the exact same situation we have now, so I'm assuming it won't be a problem. Anyway, if there's a better solution that covers all cases, then we might want to implement it here (please let me know).

manusa committed:
refactor: OpenShiftClientProducer reuses the KubernetesClient default bean

Instead of creating a new bean (which might lead to an application having two
separate KubernetesClient instances) the OpenShiftClientProducer now
relies on the default KubernetesClientProducer default bean and adapts
it.

To avoid circular dependency and multiple bean candidate issue I've used
a `@Named` annotation to qualify the default KubernetesClient bean.

However, this might be problematic in the scenario where a provides
a different base KubernetesClient bean. In this case, unless that bean
is qualified too, it won't be reused. This is the exact same situation
we have now, so I'm assuming it won't be a problem. Anyway, if there's
a better solution that covers all cases, then we might want to implement it
here (please let me know).

Signed-off-by: Marc Nuri <marc@marcnuri.com>

metacosm reviewed (approved):
Looks reasonable to me.

gsmet commented on extensions/kubernetes-client/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesClientProducer.java:
```diff
@@ -15,6 +16,7 @@ public class KubernetesClientProducer {
 
     private KubernetesClient client;
 
+    @Named("kubernetes-client")
```
Won't it break all the current consumers of `KubernetesClient`?

metacosm replied:
Hmm, actually, that's a good point. I briefly wondered about it and then thought it should be OK but I'm going to check with QOSDK.

manusa replied:
AFAIU `@Named` serves as a qualifier but not as a _"disqualifier"_, so if no other bean is provided `@Inject` should consider this bean for injection too (but maybe I'm wrong).

I understand that this test verifies this:

https://github.com/quarkusio/quarkus/blob/c9c4946859909bebf42d2fb8bfcdb8433017d230/extensions/kubernetes-client/deployment/src/test/java/io/quarkus/kubernetes/client/deployment/KubernetesClientCDITest.java#L18-L19

Anyway, maybe someone who has better understanding of CDI and ArC can confirm or deny this.

gsmet replied:
OK, if this test passes then it's all good for me.

metacosm replied:
> AFAIU `@Named` serves as a qualifier but not as a _"disqualifier"_, so if no other bean is provided `@Inject` should consider this bean for injection too (but maybe I'm wrong).
> 
> I understand that this test verifies this:
> 
> https://github.com/quarkusio/quarkus/blob/c9c4946859909bebf42d2fb8bfcdb8433017d230/extensions/kubernetes-client/deployment/src/test/java/io/quarkus/kubernetes/client/deployment/KubernetesClientCDITest.java#L18-L19
> 
> Anyway, maybe someone who has better understanding of CDI and ArC can confirm or deny this.

That was my understanding as well but I think that it doesn't stand the reality of tests :)

manusa replied:
The mentioned test passes (Not an issue with `@Named`).

The IT failed tests are related to:

https://github.com/quarkusio/quarkus/blob/ebee0a4b66d1b4dba9c51165e78d22fbf3b90d3e/integration-tests/openshift-client/src/main/java/io/quarkus/it/openshift/client/Routes.java#L15-L20

Which is yet another thing to consider. Injecting a standard `KubernetesClient` instance is now ambiguous if the OpenShiftClient extension is present.

Unless there's another option, we now need to consider removing the `@DefaultBean` annotation from the `OpenShiftClient` produced by the `OpenShiftClientProducer` and give it a very low priority so that users can still provide their own `OpenShiftClient` bean override. (this would also allow to remove the `@Named` annotation). I originally went for the `@Named` qualifier to be able to preserve the `@DefaultBean`. However, given the current scenario, I don't see any other option other than removing the `@DefaultBean` to fulfill #35383

metacosm replied:
My test only uses `@Inject KubernetesClient client` as well and fails. The "difference" maybe is that we're depending on the `quarkus-openshift-client` extension, not the `quarkus-kubernetes-client` one so maybe that changes things for CDI, since we're getting the client in an indirect way?

manusa replied:
I was trying to solve this with `@Specializes` annotation, however, I don't think it's working as expected in Quarkus.
Do you have any reference about this?

manusa replied:
> My test only uses `@Inject KubernetesClient client` as well and fails

The problem has to do with injecting `KubernetesClient` instead of `OpenShiftClient` since then it becomes ambiguous.
It seems that CDI support is limited in Quarkus and not everything seems to work. It's getting hard to solve this, especially when dealing with the `@DefaultBean` problem.

metacosm replied:
Maybe we should just let this as-is? Or request help from CDI experts?

manusa replied:
I'm trying a few more ideas, if nothing works I'll give up.

manusa replied:
https://github.com/quarkusio/quarkus/pull/35525#issuecomment-1692765728

gsmet reviewed (changes requested):
Let's sort out my question before thinking about merging this one.

manusa committed:
temp: OpenShiftClientProducer specializes KubernetesClientProducer

The right approach for this bean would be to use specialization.
https://docs.jboss.org/weld/reference/latest/en-US/html/specialization.html#_using_specialization
However, this doesn't seem to work on Quarkus.

Signed-off-by: Marc Nuri <marc@marcnuri.com>

manusa commented:
I added a temporary commit to showcase how this should look like (With some workarounds) in case someone with expertise on CDI and Quarkus can give it a look.

This commit includes two tests that verify what we want to achieve (regardless of the future of this PR, it might be interesting to keep them):
- **BeanDefaultsTest:** A user is able to inject `KubernetesClient` and `OpenShiftClient` fields.
  The injected KubernetesClient instance should be the one produced by OpenShiftClientProducer (OpenShiftClient)
- **BeanOverridesTest:** A user is able to override the **default** `OpenShiftClient` with their own instance and producer method **without** providing further configurations or annotations.
  _(requires that all of our producer methods are annotated with `@DefaultBean`. Requires our processor methods to remove one of the producers in case both extensions coexist. Discards the `@Priority` option.)_

This scenario is a great candidate for [bean specialization](https://docs.jboss.org/weld/reference/latest/en-US/html/specialization.html#_using_specialization).

However, this doesn't seem to work.

The commit has commented-out sections of how specialization should be configured (per my understanding).

As a temporary solution, for illustration purposes only, I added a delegate to show how this approach makes the test pass while solving the issue at hand.

What we need is to remove the delegate approach and rely on the super class. In addition, once this works, we should be able to remove the `KubernetesClientProcessor.registerBeanProducers` line where we conditionally add the KubernetesClientProducer so we always add it.

It would be great if someone with more expertise with Quarkus+CDI could take a look and see if there's anything wrong with the approach, or if there's actually something missing in Quarkus.

geoand commented on extensions/openshift-client/runtime/src/main/java/io/quarkus/it/openshift/client/runtime/OpenShiftClientProducer.java:
```diff
@@ -2,47 +2,37 @@
-        return client;
+    //    @Override
+    public OpenShiftClient kubernetesClient(KubernetesSerialization kubernetesSerialization, Config config) {
+        delegate = new KubernetesClientProducer();
```
Can't this be injected by CDI?

manusa replied:
Yes, that's the previous commit with the named bean approach -[full previous discussion](https://github.com/quarkusio/quarkus/pull/35525#discussion_r1304156833)-.

However, then we get the bean ambiguity when injecting a `KubernetesClient` (not OpenShift) that doesn't contain a `@Named` annotation, thus breaking the `BeanDefaultsTest`.
This can then be fixed by removing the `@DefaultBean` annotation in the `OpenShiftClientProducer`. However, then we will break the `BeanOverridesTest` (unless we further annotate the bean overrides with priority or something else).

The main problem is that if we want to preserve the frictionless approach and not introduce breaking changes to close #35383, then we need to follow the specialization approach or something along those lines.

manusa commented:
> Further, I believe the existing code already guarantees that exactly 1 KubernetesClient instance exists, because if the OpenShift capability is present, the KubernetesClientProducer is not registered as a bean -- the only KubernetesClient instance is the OpenShiftClient instance. Am I wrong there?

Yes, you're right here (I commented on the issue with the code details included: https://github.com/quarkusio/quarkus/issues/35383#issuecomment-1691841780)

>  I believe the solution with only registering one of the producer beans is fairly straightforward and elegant.

:+1: That's how it's currently implemented.



As agreed in an internal call with @metacosm and @shawkins, since specialization is not supported we can close the PR and the issue.

P.S. I added a few tests to this PR that I'll probably cherry-pick to a new PR so that we ensure the behavior of our produced beans.

metacosm commented:
Thanks, @Ladicek!
